src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Utf8.cs

```diff
@@ -0,0 +1,148 @@
+        {
+            // The empty UTF-8 string is trivially a prefix of every other string. For compat with
+            // earlier versions of the Framework we'll early-exit here before validating the
+            // 'options' argument.
+
+            if (prefix.IsEmpty)
+            {
+                return true;
+            }
+
+            if ((options & ValidIndexMaskOffFlags) == 0)
+            {
+                // Common case: caller is attempting to perform a linguistic search.
+                // Pass the flags down to NLS or ICU unless we're running in invariant
+                // mode, at which point we normalize the flags to Ordinal[IgnoreCase].
+
+                if (!GlobalizationMode.Invariant)
+                {
+                    return StartsWithCoreUtf8(source, prefix, options);
+                }
+
+                if ((options & CompareOptions.IgnoreCase) == 0)
+                {
+                    return source.StartsWith(prefix);
+                }
+
+                return source.StartsWithOrdinalIgnoreCaseUtf8(prefix);
+            }
+            else
+            {
+                // Less common case: caller is attempting to perform non-linguistic comparison,
+                // or an invalid combination of flags was supplied.
+
+                if (options == CompareOptions.Ordinal)
+                {
+                    return source.StartsWith(prefix);
+                }
+
+                if (options == CompareOptions.OrdinalIgnoreCase)
+                {
+                    return source.StartsWithOrdinalIgnoreCaseUtf8(prefix);
+                }
+
+                ThrowCompareOptionsCheckFailed(options);
+
+                return false; // make the compiler happy;
+            }
+        }
+
+        private unsafe bool StartsWithCoreUtf8(ReadOnlySpan<byte> source, ReadOnlySpan<byte> prefix, CompareOptions options)
+        {
+            // NLS/ICU doesn't provide native UTF-8 support so we need to convert to UTF-16 and compare that way
+
+            // Convert source using stackalloc for <= 256 characters and ArrayPool otherwise
+
+            char[]? sourceUtf16Array;
+            scoped Span<char> sourceUtf16;
+            int sourceMaxCharCount = Encoding.UTF8.GetMaxCharCount(source.Length);
+
+            if (sourceMaxCharCount <= 256)
+            {
+                sourceUtf16Array = null;
+                sourceUtf16 = stackalloc char[512];
+            }
+            else
+            {
+                sourceUtf16Array = ArrayPool<char>.Shared.Rent(sourceMaxCharCount);
+                sourceUtf16 = sourceUtf16Array.AsSpan(0, sourceMaxCharCount);
+            }
+
+            OperationStatus sourceStatus = Utf8.ToUtf16(source, sourceUtf16, out _, out int sourceUtf16Length, replaceInvalidSequences: false);
+
+            if (sourceStatus != OperationStatus.Done)
+            {
+                return false;
```

Comment:
This gets a bit complicated. The code as-is definitively returns "not a prefix match!" - even though the inputs are invalid to the point that the question is nonsensical. It's a bit like asking "is 'dog' less than 'cat'?" It's not true and it's not false. It's just... huh?

Throwing a wrench into this even more is that ICU actually _does_ have specialized handling for invalid UTF-16 sequences. (Basically, it treats them as opaque chars that don't match anything except for themselves.) So this means that, e.g., the sequence `"\uD800\uD801".StartsWith("\uD800")` would return _true_, even though both sequences are ill-formed. But in the logic you have here, `"\xFF\xFE"u8.CultureAwareStartsWith("\xFF"u8)` would return _false_.

Recommendation: Invalid UTF-8 sequences should throw (not return false) or should utilize existing ICU handling for invalid UTF-16 sequences.

The way I worked around this in the UTF-8 prototype was to use something akin to WTF-16 for the comparison. Basically, every time an invalid UTF-8 byte is encountered, append an invalid UTF-16 code point to the string we're building. Any invalid UTF-8 byte `0x??` would map to the char `U+DF??`. So for instance, the invalid byte `0x80` would map to the invalid char `0xDF80`, the invalid byte `0xC0` would map to the invalid char `U+DFC0`, etc. The end result is that you're building up something that is _mostly_ UTF-16, with the invalid sequences specifically chosen not to conflict with one another, which allows ICU to handle the input string appropriately.

This does not have the same algorithmic complexity constraints that I call out in `IUtf8SpanParsable`, since ICU's implementation for culture-aware prefix matching is `O(needle.Length * haystack.Length)` with no early-exit optimization. It's lamentable, but it is what it is. So that means the transcoding logic is upping some constant factor but isn't otherwise changing the algorithmic complexity itself.

(NLS, I believe, _does_ have an early exit consideration. But that should be enough of an edge case that I'm willing to hold my nose for any algorithmic complexity changes we introduce.)